composer/core/state.py

```diff
@@ -359,6 +359,9 @@ def __init__(
 
         self.train_metrics: Dict[str, Metric] = {}
         self.eval_metrics: Dict[str, Dict[str, Metric]] = {}
+        self.train_metric_values: Dict[str, float] = {}
+        self.eval_metric_values: Dict[str, float] = {}
+        self.total_loss_dict: Dict[str, float] = {}
```

Comment:
whats this for?

Reply:
it is so we can store the most recent values of the metrics, so that they can be logged upon Event instead of being logged during a `log_metrics` call. The reason we do this is so we can have access to State to look up the batch and epoch. See `ConsoleLogger`